include/veriblock/abfi/bitcoin/serialize.hpp

```diff
@@ -0,0 +1,1016 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2018 The Bitcoin Core developers
+// Copyright (c) 2019-2021 Xenios SEZC
+// https://www.veriblock.org
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SERIALIZE_H
+#define BITCOIN_SERIALIZE_H
+
+#include <stdint.h>
+#include <string.h>
+
+#include <algorithm>
+#include <ios>
+#include <limits>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "veriblock/pop/blob.hpp"
+#include "veriblock/pop/read_stream.hpp"
+#include "veriblock/pop/slice.hpp"
+#include "veriblock/pop/write_stream.hpp"
+
+namespace btc {
+
+static const unsigned int MAX_SIZE = 0x02000000;
+
+/**
+ * Dummy data type to identify deserializing constructors.
+ *
+ * By convention, a constructor of a type T with signature
+ *
+ *   template <typename Stream> T::T(deserialize_type, Stream& s)
+ *
+ * is a deserializing constructor, which builds the type by
+ * deserializing it from s. If T contains const fields, this
+ * is likely the only way to do so.
+ */
+struct deserialize_type {};
+constexpr deserialize_type deserialize{};
+
+/**
+ * Used to bypass the rule against non-const reference to temporary
+ * where it makes sense with wrappers.
+ */
+template <typename T>
+inline T& REF(const T& val) {
+  return const_cast<T&>(val);
+}
+
+/**
+ * Used to acquire a non-const pointer "this" to generate bodies
+ * of const serialization operations from a template
+ */
+template <typename T>
+inline T* NCONST_PTR(const T* val) {
+  return const_cast<T*>(val);
+}
+
+//! Safely convert odd char pointer types to standard ones.
+inline char* CharCast(char* c) { return c; }
+inline char* CharCast(unsigned char* c) { return (char*)c; }
+inline const char* CharCast(const char* c) { return c; }
+inline const char* CharCast(const unsigned char* c) { return (const char*)c; }
+
+/*
+ * Lowest-level serialization and conversion.
+ * @note Sizes of these types are verified in the tests
+ */
+template <typename Stream>
+inline void ser_writedata8(Stream& s, uint8_t obj) {
+  s.write((char*)&obj, 1);
+}
+template <typename Stream>
+inline void ser_writedata16(Stream& s, uint16_t obj) {
+  obj = htole16(obj);
+  s.write((char*)&obj, 2);
```

Comment:
if we use WriteStream here, we don't need htole16, because stream has .writeLE<uint16_t>(obj)